### PR TITLE
Refactor blazemod [RDY]

### PR DESCRIPTION
**Pull request by mugling**
Cleanup of #17417 using #17431

### CODE REVIEW

**Comment by mugling** [comment]
Done

**Comment by codemime** [comment]
The PR is damn huge.

**Comment by mugling** [comment]
> The PR is damn huge.

Yes, no way around that - reformatting only adds ~30% versus the parent PR (#17417)

**Comment by Coolthulhu** [comment]
Is it tested or only reformatted?

File: data/mods/blazemod/blaze_ammo.json [hunk 1]
```diff
@@ -1,412 +1,386 @@
...
+    "bashing": 8,
+    "material": "iron",
+    "symbol": "*",
+    "color": "red",
+    "count": 1,
+    "stack_size": 1,
+    "ammo_type": "hbolt",
+    "damage": 14,
+    "effects": [ "FLAME", "NEVER_MISFIRES", "COOKOFF" ],
+    "flags": [ "" ]
```
Comment (Coolthulhu) [inline data/mods/blazemod/blaze_ammo.json]: This line is horrible. Can linter detect and remove those too?

File: data/mods/blazemod/blaze_ammo.json [hunk 2]
```diff
@@ -1,412 +1,386 @@
...
+    "description": "An IED composed of a container filled with a payload material affixed to the end of a large ballista bolt.  Intended to be fired from a specialized launcher.  This one is designed to douse a small area with deadly flames.",
+    "weight": 3816,
+    "volume": 1,
+    "price": 600,
+    "bashing": 8,
+    "material": "iron",
+    "symbol": "*",
+    "color": "red",
+    "count": 1,
+    "stack_size": 1,
```
Comment (Coolthulhu) [inline data/mods/blazemod/blaze_ammo.json]: Aren't `"count": 1` and `"stack_size": 1` redundant?

File: data/mods/blazemod/blaze_ammo.json [hunk 3]
```diff
@@ -1,412 +1,386 @@
...
+    "ammo_type": "hbolt",
+    "damage": 14,
+    "effects": [ "NAPALM", "NEVER_MISFIRES", "COOKOFF" ],
+    "flags": [ "" ]
+  },
```
Comment (Coolthulhu) [inline data/mods/blazemod/blaze_ammo.json]: This entire line of items is a giant block of horrible copy+paste. Would benefit from some "similar json detector" system that would replace sets of definitions with `copy-from` entry.
Extending linter to do that would probably be too much work (it's too specialized), but I could write a custom script just to do that for this PR.

File: data/mods/blazemod/blaze_ammo.json (hunk 1 above)
Comment (mugling) [inline data/mods/blazemod/blaze_ammo.json]: > This line is horrible. Can linter detect and remove those too?

I could extend it to provide a `NOEMPTY` output flag

File: data/mods/blazemod/blaze_ammo.json (hunk 2 above)
Comment (mugling) [inline data/mods/blazemod/blaze_ammo.json]: Yes, but that is out of scope of the linter - it's valid JSON, a valid field and sensibly formatted.

Strict mode (#17722 etc) could detect this but cannot be enabled by default for mods due to the false positives that would be reported due to the variation in load order. One option would be a command-line switch that selectively tested a mod in strict mode and output a non-compliance log?

**Comment by mugling** [comment]
> Is it tested or only reformatted?

It is only reformatted. I haven't addressed balance or anything like that.

**Comment by Coolthulhu** [comment]
> I haven't addressed balance or anything like that.

Balance can be off (most likely is), but someone will have to bite the bullet and go through all of this crap to check if it actually works or spews debugmsgs on every attempt.

Or we can just say "fuck it" and merge it with the assumption that mod content can be full retard. But then we'll have to maintain it.

> Yes, but that is out of scope of the linter - it's valid JSON, a valid field and sensibly formatted.

Could you run a separate script for that? Some simple `jq` equivalent of:
```
if( entry["type"] == "ammo" && entry["count"] == 1 ) entry.erase( "count" );
```

and ditto for "stack_size".

The PR is unreviewably huge and ugly. Some quick and dirty solutions may be necessary to trim it to manageable size.

**Comment by mugling** [comment]
Yes, some `jq`-fu could be in order

> Or we can just say "fuck it" and merge it with the assumption that mod content can be full retard.

It's better by far than the status quo and to be honest I can only justify further effort in this PR in that context.

> But then we'll have to maintain it.

This is effectively a one-time deal. We should do everything possible to avoid getting to this situation again. Maintenance should be more practical given the lack of JSON errors and the author becomes responsible for any fixes internal to the mod.

Summarily if I run this through `jq` and possibly `sed` and get no load errors are you agreeable to a merge?

**Comment by Coolthulhu** [comment]
> Summarily if I run this through jq and possibly sed and get no load errors are you agreeable to a merge?

I'd still have objections. This needs automated testing.
I can write a unit test specifically for it.
Since we're essentially re-mainlining a mod, it should at the very least work. Doesn't need to work well, but should work.

I see a unit test like this:
* Load blazemod
* For every single turret in the mod:
* Spawn single-frame vehicle
* Install full storage battery and the turret on the vehicle
* Find a holder for the default ammo of this turret
* If it exists, install a holder and fill it to the brim with default ammo
* If it doesn't, try to fill the turret instead
* Fire the turret at random tile within its range
* Wait a turn
* If no debugmsgs happen, continue, otherwise print that this turret is wonky and mark the test as failure

I assume many turrets will fail it, so whatever commands you're using to clean this up you should save in a script. We'll have to work with blaze to make the mod at least work before it's really mergeable, so we'll most likely have to clean it up more than once.

**Comment by mugling** [comment]
> so we'll most likely have to clean it up more than once.

Could we merge this PR and then proceed with incremental cleanup from there?

**Comment by Coolthulhu** [comment]
> Could we merge this PR and then proceed with incremental cleanup from there?

It technically breaks a working mainlined mod, which would be a regression.
Incremental progress has to wait until it is unbroken.
Fixes for ugliness, unreviewability, horrible duplicate entry everything etc. can then be incremental.

**Comment by mugling** [comment]
What's the minimum list then and we can work on that?

**Comment by Coolthulhu** [comment]
Minimum list of what? Things that should work?

I'm writing a unit test like the one I described above. When that passes, it will be good to go.

**Comment by mugling** [comment]
Sounds good

**Comment by BorkBorkGoesTheCode** [comment]
Do you need help?

**Comment by Zireael07** [comment]
When is this coming?

**Comment by mugling** [comment]
Waiting on #17746

**Comment by BorkBorkGoesTheCode** [comment]
Are you still interested in merging this?

**Comment by mugling** [comment]
Needs a rebase then some fixes to pass the unit tests but we should be otherwise ready. @Coolthulhu anything else?

**Comment by Coolthulhu** [comment]
Once it passes the firing test and is mergeable, it should be better than the broken version we have in master.

**Comment by mugling** [comment]
This is very difficult to rebase

**Comment by mugling** [comment]
61631c4 fixes a bug whereby specifying `terrain:symbol` or `terrain:color` as an array with one element results in a cryptic error message. The new version supports loading a string, an array with a single element or an array of four elements. Anything else yields a descriptive error message.

Almost ready:

- [X] ./cataclysm-tiles --check-mods blazemod
- [X] gmake json-lint
- [ ] ./tests/cata_test --mods=blazemod [gun]

**Comment by mugling** [comment]
Passes all automated tests and considerably easier to work on than before. No doubt the mod contains other issues but I'm going to merge this as per the above discussion.
